ENH: structured datatype safety checks

ahaldane opened the pull request:
This PR tries to solve some issues related structured datatypes, especially view/getfield/setfield. Previously views of structured arrays containing objects were completely disabled.  This commit adds more lenient check for whether an object-array view is allowed, and adds similar checks to getfield/setfield

(This PR has been edited around 3/4/15. I originally planned other changes which I will submit as separate PRs)

I believe this PR solves issues https://github.com/numpy/numpy/issues/2599, https://github.com/numpy/numpy/issues/3256, https://github.com/numpy/numpy/issues/3286, https://github.com/numpy/numpy/issues/2346, https://github.com/numpy/numpy/issues/3253. Issue https://github.com/numpy/numpy/issues/5081 is also now partially fixed, but requires further thought (please take a look at the example there). It also fixes the new scipy issue https://github.com/scipy/scipy/issues/4589.

Only one test needed modification to pass: `test_objectarray_setfield` checks that `setfield` with objects is disallowed, and expects a `RuntimeError` but I give it `TypeError`.
### Summary

Previously views of structured arrays containing objects were completely disabled, which was a problem since views are needed in a few important cases such as when indexing multiple fields of a structured array, or in `__getitem__` in record arrays.

With help from @jaimefrio I now have functions that carefully check whether a view into array would interpret non-Object memory as an object, or vice versa, and allow the view if there is no problem. Big thanks to Jaime for help, code, and explanations on this. I think the checks are compatible with his PR https://github.com/numpy/numpy/pull/5508. I check that 1) No memory that is not an object will be interpreted as a object, 2) No memory containing an object will be interpreted as an arbitrary type. Both cases can cause segfaults.

I also added similar checks to `PyArray_GetField` and `PyArray_SetField`. Previsouly, the user could view arbitrary data as a python Object, and it was fairly easy to make cases where a segfault would happen.

Note, however, that there are still "unsafe" operations (wrt Objects) in numpy which I hope to examine in future PRs, when I have time.
### How to keep track of objects

There is a design choice here worth discussing. As part of the check for objects when taking a view, we need to go though the new dtypes's fields and make sure none of them overlap with memory containing an object in the old dtype (unless the field is also an object). Consider this potential view:

```
>>> basearr = np.array([(1,2,3),(4,5,6)], 
...                    dtype=[('f0', 'i8'), ('f1', 'O'), ('f2', 'i8')])
>>> varr = basearr.view({'names': ['f0','f2'], 
...                      'formats': ['i8','i8'], 
...                      'offsets': [0,16]})
>>> v2arr = varr.view({'names': ['f0'], 
...                    'formats': ['i8'], 
...                    'offsets': [8]})  
```

The problem would be that writing to v2arr will overwrite the Object in basearr. We need to prevent this, but the challenge is that `varr` doesn't "know" that there is an object at position 8. This raises the question: How do we figure out where the objects are in memory to prevent them from being viewed?

The strategy I have taken in this PR is to only allow views into fields "known" to the parent array. In the example above, the `v2arr` view would be disallowed, and only views of `varr` into offsets 0 and 16 would be allowed. (specifically, views into bytes 0-8 and 16-24). The downside is that many views are no longer "reversible", since once you take a view that "forgets" what was at an offset, you can never view the memory at that offset again in further views. While this constraint only really has to apply to object arrays (as tested with `NPY_ITEM_HASOBJECT`, which I think is inherited in all views) I think for consistency it should apply to all structured arrays.. That is, all 'partial' views (views that have less fields than the original array) would be irreversible.

An alternate strategy is to too look up all the object positions in the "base" object of the array, `varr.base` in this case. I assume that the "base" array (if it is an array) contains the list of all object offsets, which I use to check for view overlaps. I implemented this in a separate branch (partial version [here](https://github.com/ahaldane/numpy/tree/view_objects_base)), but it was difficult because `arr.base` does not always lead to an array object. For example it may be a buffer, in which case I disallow views containing object types, but that's not a big problem because users aren't allowed to create object arrays from buffers anyway. Another problem is (void) scalar structured types, as they do not list their "base" object even after being converted to a 0-d array. To fix this I made void scalar types' base attribute point to the original array. However, there are still some cases where I cannot get to the original array - for example `nditer` buffers the source array, which causes about 5 unit tests to fail, which I have not solved yet (but I could try). The advantage would be that views are reversible.

The irreversible view strategy seems better because it touches less parts, has fewer edge cases, and seems to me less likely to go wrong. It also works right now. I also have not imagined any scenarios where inability to reverse a view would be a big problem, and no related unit tests fail.

Note that even if the 'base array' strategy is not good, perhaps it's still a good idea to make void scalars "base" attribute point to the original array, as another PR? I admit I am not totally clear on the true purposes of "base" objects.

In this PR I also made some small tweaks to recarrays: I don't think they need to implement their own `view` function as a result of my last PRs.

ahaldane committed:
ENH: structured datatype safety checks

Previously views of structured arrays containing objects were completely
disabled.  This commit adds more lenient check for whether an object-array
view is allowed, and adds similar checks to getfield/setfield

Fixes #2346. Fixes #3256. Fixes #2599. Fixes #3253. Fixes #3286.
Fixes #5762.

ahaldane commented:
All right, I think this is done. I'd appreciate any comments or suggestions!

jaimefrio commented on numpy/core/_internal.py:
```diff
@@ -305,6 +305,106 @@ def _index_fields(ary, fields):
+            count = 1
+            for dim in dtype.shape:
+                count *= dim
+            fields.extend((typ,off + dtype.base.itemsize*j)
```
space after comma

jaimefrio commented on numpy/core/_internal.py:
```diff
@@ -305,6 +305,106 @@ def _index_fields(ary, fields):
+            for dim in dtype.shape:
+                count *= dim
+            fields.extend((typ,off + dtype.base.itemsize*j)
+                                              for j in range(count)
```
This is probably better aligned with the beginning of `(typ, off...`

jaimefrio commented on numpy/core/_internal.py:
```diff
@@ -305,6 +305,106 @@ def _index_fields(ary, fields):
+                count *= dim
+            fields.extend((typ,off + dtype.base.itemsize*j)
+                                              for j in range(count)
+                                              for typ,off in sub_offsets)
```
space after comma

jaimefrio commented on numpy/core/_internal.py:
```diff
@@ -305,6 +305,106 @@ def _index_fields(ary, fields):
+
+    #first go byte by byte and check we do not access bytes not in old_fields
+    new_bytes = set()
+    for tp,off in new_fields:
```
Another (missing) space after comma, I see many more looking down, won't point any other out, but please have them all fixed.

jaimefrio commented on numpy/core/_internal.py:
```diff
@@ -305,6 +305,106 @@ def _index_fields(ary, fields):
+    from .numerictypes import object_
+    from .multiarray import dtype
+
+    #first go byte by byte and check we do not access bytes not in old_fields
```
There are smarter ways of going about this check, but we probably shouldn't let micro-optimizations distract us from the larger goal.

jaimefrio commented on numpy/core/_internal.py:
```diff
@@ -305,6 +305,106 @@ def _index_fields(ary, fields):
+    old_bytes = set()
+    for tp,off in old_fields:
+        old_bytes.update(set(range(off, off+tp.itemsize)))
+    if len(new_bytes.difference(old_bytes)) != 0:
```
PEP 8 says you should check this as `if new_bytes.difference(old_bytes):` taking advantage of empty sets being `False`. It actually says so for sequences, but I guess it also applies to sets.

jaimefrio commented on numpy/core/_internal.py:
```diff
@@ -305,6 +305,106 @@ def _index_fields(ary, fields):
+        raise TypeError("view would access data parent array doesn't own")
+
+    #next check that we do not interpret non-Objects as Objects, and vv
+    obj_offsets = [off for tp,off in old_fields if tp == object_]
```
Isn't the correct check here `if tp is object_`? I'm just raising the question, not really sure.

jaimefrio commented on numpy/core/_internal.py:
```diff
@@ -305,6 +305,106 @@ def _index_fields(ary, fields):
+            if fld_offset not in obj_offsets:
+                raise TypeError("cannot view non-Object data as Object type")
+        else:
+            # next check we does not create non-object views
```
do, not does

jaimefrio commented on numpy/core/_internal.py:
```diff
@@ -305,6 +305,106 @@ def _index_fields(ary, fields):
+    old_fields = get_all_field_offests(oldtype)
+    old_size = oldtype.itemsize
+
+    #if the itemsizes are not equal, we need to check that all the
```
space after `#`

jaimefrio commented:
Thanks for putting this together, @ahaldane. There has obviously been a lot of hard thinking going on!

I will go on reviewing later, just two things for now, aside from all the style nitpicks:

 1. It would be nice to add tests for the new functionality in `_internal.py`, without relying on other stuff. For instance, we will not be able to check the tiling of views unless we merge something like #5508, but it would be nice to know that it works properly beforehand.
 2. I also think that the overlap checks could be made smarter, probably by sorting the fields by offset and fusing adjacent non-objects together, then checking them in order. That can be left for another PR, especially if we add proper tests for those functions now.
 3. I'm not sure if you already had, but if you haven't, it would be a good idea to ping the list on the specific topic of "should views into structured arrays be reversible?"

ahaldane commented:
Thanks for looking over it @jaimefrio.

I think I've fixed the style issues, and I also made a few other small changes. I made it a second commit for clarity on what's changing.

I'll look into adding some extra tests, and I agree it might be possible to make the overlap checks smarter. I'll also ping the list.

jaimefrio commented on numpy/core/src/multiarray/getset.c:
```diff
@@ -449,15 +450,21 @@ array_descr_set(PyArrayObject *self, PyObject *arg)
+    /* check that we are not reinterpreting memory containing Objects */
+    _numpy_internal = PyImport_ImportModule("numpy.core._internal");
+    if (_numpy_internal == NULL) {
+		Py_DECREF(newtype);
```
We only indent more than 4 spaces if the `if` condition spans more than 1 line.

jaimefrio commented on numpy/core/src/multiarray/getset.c:
```diff
@@ -449,15 +450,21 @@ array_descr_set(PyArrayObject *self, PyObject *arg)
+                               PyArray_DESCR(self), newtype);
+    Py_DECREF(_numpy_internal);
+    if (safe == NULL) {
+		Py_DECREF(newtype);
```
Same indentation issue here.

jaimefrio commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -418,18 +425,7 @@ PyArray_SetField(PyArrayObject *self, PyArray_Descr *dtype,
-                           dtype, PyArray_NDIM(self), PyArray_DIMS(self),
-                           PyArray_STRIDES(self), PyArray_BYTES(self) + offset,
-                           PyArray_FLAGS(self), (PyObject *)self);
+    ret = PyArray_GetField(self, dtype, offset);
```
Interesting approach. Perhaps it warrants an explanatory comment that the return is a view, not a copy?

jaimefrio commented:
I only had some minor style corrections on the rest of your code.

What I would like to see to confidently merge it is a good set of tests, but the code itself looks good to go.

ahaldane commented:
Indentation fixed (was because my editor didn't expand tabs for some reason), added comment about getfield, and posted to list.

I also added some tests. Do you think they are comprehensive enough, or do you have anything other ideas to test?

ahaldane commented:
By the way, something else caught my eye: I had made Setfield call Getfield because the setup code inside both functions seemed identical. But actually there is a difference: Getfield masks the array flags with `~NPY_ARRAY_F_CONTIGUOUS` but setfield didn't. Why is that?

I looked into it, and the mask is there to fix a bug related to subarrays of void scalars. However, I think it's a bug from a long time ago which is more or less irrelevant after commits from summer 2014. So we probably could remove the mask, but it also seems harmless to leave it in. The fact that setfield didn't mask too is probably another bug which this PR fixes. If you would like more details I am happy to explain (it's a bit of a ride).

charris commented:
Rejigger tests by closing and opening.

ahaldane commented:
(rebased)

charris commented:
I think we need to get this in. @jaimefrio How does this look to you now.

charris commented on numpy/core/records.py:
```diff
@@ -525,22 +525,6 @@ def field(self, attr, val=None):
         else:
             return self.setfield(val, *res)
 
-    def view(self, dtype=None, type=None):
```
Structured types now have the needed functionality?

charris commented:
LGTM on (shallow) study. Note that all the existing functions in `_internal.py` are private, might be could to continue that practice. The new functions in `_internal.py` need docstrings. None of the old functions have docstrings, but that is because they are relics from a time before documentation was invented.

jaimefrio commented:
I started re-reviewing it last night, but it's great bedtime reading, and I didn't make it all the way through. I had a couple of existential doubts while looking through the code, will try to word when I continue tonight.

charris commented:
@ahaldane You need separate `Fixes #2346. Fixes #3256. Fixes #2599, Fixes #3253. Fixes #3286`. to automatically close multiple issues.

jaimefrio commented on numpy/core/tests/test_records.py:
```diff
@@ -206,6 +206,15 @@ def test_pickle_2(self):
 
+    def test_objview_record(self):
+        # https://github.com/numpy/numpy/issues/2599
+        dt = np.dtype([('foo', 'i8'), ('bar', 'O8')])
```
We have deprecated `'O8'`, better to use `'O'`, or this test will break pretty soon.

jaimefrio commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -808,6 +808,14 @@ def test_casting(self):
+        a[['a', 'b']] # TypeError?
+
+        # https://github.com/numpy/numpy/issues/3253
+        dat2 = np.zeros(3, [('A', 'i'), ('B', '|O8')])
```
Better `'|O'` here also.

charris commented:
@ahaldane Could you make the fixes, squash the commits, and rebase this? Might rebase #4652 also.

ahaldane commented:
Sure, I'll do it soon.

ahaldane commented:
All right, fixed+rebased+squashed. 

I got rid of the changes to records.py since they were unrelated to the main changes here (I'll put them in another PR for recarrays I want to do at some point).

I'm curious what the "existential doubts" Jaime mentioned are :). Maybe there are potential issues that didn't occur to me?

charris commented:
Haven't yet looked through the code myself, but we do need the record fix for scipy before 1.10. Does this still do that?

ahaldane commented:
You mean #5762? Yes, I checked that is fixed.

Also I just saw that https://github.com/numpy/numpy/pull/5921 already made the recarray change I just removed from this PR. Cool.

charris commented on numpy/core/_internal.py:
```diff
@@ -305,6 +305,115 @@ def _index_fields(ary, fields):
     return array(view, dtype=copy_dtype, copy=True)
 
+def _get_all_field_offsets(dtype, base_offset=0):
+    """ Returns the types and offsets of all fields in a dtype,
```
Docstrings should be formatted in the standard numpy way.

charris commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -358,15 +358,22 @@ NPY_NO_EXPORT PyObject *
 PyArray_GetField(PyArrayObject *self, PyArray_Descr *typed, int offset)
 {
     PyObject *ret = NULL;
+    PyObject *_numpy_internal, *safe;
```
This is usually done using a static variable initialized on first call. For instance
```
    static PyObject *matmul = NULL;

    if (matmul == NULL) {
        PyObject *module, *dict;

        module = PyImport_ImportModule("numpy.core.multiarray");
        dict = PyModule_GetDict(module);
        matmul = PyDict_GetItemString(dict, "matmul");
        Py_INCREF(matmul);
        Py_DECREF(module);
    }
```

charris replied:
We should probably make a macro for that at some point.

charris commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -358,15 +358,22 @@ NPY_NO_EXPORT PyObject *
+        return NULL;
+    }
+    /* only returns True or raises */
+    safe = PyObject_CallMethod(_numpy_internal,
```
This then becomes `PyObject_CallFunction`.

charris commented on numpy/core/src/multiarray/methods.c:
```diff
@@ -358,15 +358,22 @@ NPY_NO_EXPORT PyObject *
+    /* only returns True or raises */
+    safe = PyObject_CallMethod(_numpy_internal,
+            "_getfield_is_safe", "OOi", PyArray_DESCR(self), typed, offset);
+    Py_DECREF(_numpy_internal);
```
And this goes away.

ahaldane commented:
Docs updated. 

I'll change to static variable initialization in a bit. Although, maybe that's better left for another PR? There are about 12 other places that could use static initialization but don't (I copied from one of them).

charris commented:
Yeah, I just put up #5940 to provide a function to do the caching. If it goes in I'll fix up the  imports.

jaimefrio commented:
If we leave the static allocation for another PR, is this the ready to go?

ahaldane commented:
Static allocation is now in, using #5940. It looks like I don't need to set the error string if npy_cache_pyfunc fails, since it will already be set. If that's fine, then I think this is done.

charris commented:
@jaimefrio My concerns have been taken care of.

jaimefrio commented:
It shall then be merged, @charris ;-)

jaimefrio commented:
Thanks, @ahaldane!

ahaldane commented:
Thanks to both of you too!
